Extend apiv3 blueprint for user (un-)locking

oliverguenther opened the pull request:
With OmniAuth, we now have external services authenticate
users in an asnychronous manner. With this change, proper means
to manage the status based from these services become a necessity.
Particularly, deleting and locking users based on outer events are
a requirement for timely account administration.

As the API(v3) already provides the functionality to remove user
accounts, this commit suggests an extension to provide locking
and unlocking user accounts through the API.

Given the current api organization, two options came to mind:
1. PATCHing the user resource `/api/v3/users/{id}` with a flag the corresponds to the lock status
2. Adding a resource  `/api/v3/users/{id}/lock`, for which `GET` returns the current status, `POST` performs the lock and `DELETE` removes it.

This PR suggests the latter, as I believe the former is ambiguous (and PATCH should be reserved for updating a subset of simple user attributes). 

I apologize for any blatant errors in the blueprint, the apib schema was unknown to me until now :fearful: 

The discussion leading up to this PR: https://community.openproject.org/topics/3706
The relevant WP: https://community.openproject.org/work_packages/18122

oliverguenther committed:
Extend apiv3 blueprint for user (un-)locking

With OmniAuth, we now have external services authenticate
users in an asnychronous manner. With this change, proper means
to manage the status based from these services become a necessity.
Particularly, deleting and locking users based on outer events are
a requirement for timely account administration.

As the API(v3) already provides the functionality to remove user
accounts, this commit suggests an extension to provide locking
and unlocking user accounts through the API.

TeatroIO commented:
I've prepared a stage. <a href='http://oliverguenther-feature-apiv3-user-lockunlock.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Click to open.</a>

oliverguenther committed:
Added empty body to please snowcrash

ulferts commented:
Thanks for the PR @oliverguenther. Can you please take a look at it @NobodysNightmare?

NobodysNightmare commented:
Hello, I had a first look at your PR and like it quite much.

I agree that the _normal_ `PATCH` does not sound very apropriate for that use case, a separate endpoint feels better here. I also like that you determine if the lock is being set or removed using the HTTP verb (`DELETE` vs. `POST`).

_Side note_: This is something we should also consider for the **star** endpoint of the query, which currently provides two distinct resources `star` and `unstar` (not to mention that those are for some reason called with a `PATCH` request).

I am not yet sure how I feel about the `GET` part of your proposal. On the one hand it seems very consequent to me, because if there is something that I can `DELETE`, then I should also be able to `GET` it. On the other hand I feel that I would have simply put the lock state into the user object.
I am not very decided on that topic yet, it is more like a vague feeling, you might have an opinion or further thoughts on that too :)

After all **thanks for your proposal**, it looks good to me.
However, I will have to think and talk about some bits and pieces before merging it. I'll tell you if I see any need for changes.

If I may ask: Would you also be able/willing to implement the specified functionality after this specification change is accepted? :smile:

NobodysNightmare commented on doc/apiv3-documentation.apib:
```diff
@@ -2041,6 +2041,134 @@ Permanently deletes the specified user account.
+                    }
+                },
+                "_type": "Lock Status",
+                "locked": true,
```
minor syntactic error: superfluous comma

oliverguenther commented:
Thanks for your remarks! I was wondering whether the lock status could simply be integrated into the existing `GET /users/<id>`. However for the sake of the PR, I isolated the feature in that separate group. I'm fine with both and will gladly modify the PR to match your intended design.

I will provide a PR implementing the feature once the design is fixed.

NobodysNightmare commented on doc/apiv3-documentation.apib:
```diff
@@ -2041,6 +2041,134 @@ Permanently deletes the specified user account.
+            {
+                "_type": "Error",
+                "errorIdentifier": "urn:openproject-org:api:v3:errors:MissingPermission",
+                "message": "You are not allowed to lock the account of this user."
```
**un** lock

ulferts commented on doc/apiv3-documentation.apib:
```diff
@@ -2041,6 +2041,134 @@ Permanently deletes the specified user account.
+                        "title": "John Sheppard - admin"
+                    }
+                },
+                "_type": "Lock Status",
```
We are using camel case for the types, e.g. "WorkPackage"

NobodysNightmare commented:
I just noticed that the user object is currently lacking an actions section that describes the action links a user object might contain (including the actions that you just added).
I've prepared one for you to include below `# Group Users`:

```
## Actions:
| Link                | Description                                                          | Condition                                 |
|:-------------------:| -------------------------------------------------------------------- | ----------------------------------------- |
| lock                | Restrict the user from logging in and performing any actions         | not locked; **Permission**: Administrator |
| unlock              | Allow a locked user to login and act again                           | locked; **Permission**: Administrator     |
```

The examples containing user objects also need to be extended to contain those links (I am just referencing a links section here):

``` JSON
                "_links": {
                    "self": {
                        "href": "/api/v3/users/1",
                        "title": "John Sheppard - j.sheppard"
                    },
                    "lock": {
                        "href": "/api/v3/users/1/lock",
                        "method": "POST"
                    }
                }
```

_(I noticed just now, that I forgot to include the action `delete` in the first place, I will probably add that one right after merging your PR :smiley: )_

ulferts commented:
The proposal looks consistent and straight forward. I have some points which I ask you to understand not as meant to sabotage the proposal but are questions that have arisen while reading:
- The status of the user is not something that is binary (locked/unlocked). Curently, "registered" exists at the least and why not taking into account additional states. That raises the question of whether all transitions between those states are explicit. I guess they are but that is one point we need to talk about:
  - active + "POST lock" -> locked
  - registered + "POST lock" -> locked
  - locked + "DELETE lock" -> active
- What is to be returned, when the request causes a noop, e.g. locked account + "POST lock". We can probably make this idempotent.
- What do we do with the other states, they will probably need separate endpoints then as well. At least for "activate". This would enable activating registered accounts via the API. We will then however have to define what happens with locked accounts. They should probably be unlocked, so "POST activate" would be the same as "DELETE lock". How about the opposite though ("DELETE activate")? For "registered", having a separate endpoint seems strange to me but would be consistent.
- If we raise the status to it's somewhat separate resource, although it is not explicitly mentioned as status, we have to consider what to return with the user object (see @NobodysNightmare's remark). We will probably have to embed this as it might become hard otherwise to handle the index action. How will a client know otherwise, which users are active/locked/registered.
- Similar to the one before: How do we handle filtering on the index action if lock/active/registered is it's own resource.

oliverguenther committed:
Minor changes regarding the lock user definition

- Added ``lock`` , ``delete`` links to the user model
- Removed superfluous comma
- Change ``_type``to CamelCase

oliverguenther commented:
I've corrected all remarks up until the latest comment (added the `delete` link as well for completeness).

Jens, concerning your points - These comments reflect our intended use of the api extension:

**Status transtitions**
If I'm not mistaken, locking a registered user from the admin panel is not possible. Also, Simply locking and unlocking ( to status active ) would skip the activation procedure (e.g., activation notification mail).

I see two options:
1. Have the API be explicit about the transitions it supports: Return a 400 if the user state is not active (endpoint /lock), and locked (/unlock) to avoid these side effects.
2. Allow the transition (registered --> locked) in both API and frontend, with two issues remaining for discussion: activation/notification after unlocking + should user status after unlocking be active or registered?

**Idempotent transitions**

From our API client perspective, idempotent transitions are fine.
If our service has the intention to lock a user account only to find the account already locked through some event (e.g., manually by an admin) , it should simply carry on. The same argumentation holds for the inverse transition.
The API could output a warning message instead of 204 in that context, but I can't imagine a case where that would be useful.

**(De-)Activation transitions**
I agree that an API endpoint for user activation would be useful, especially with the integration of the other transitions.
However, while I can see the point of consistency, using an endpoint `DELETE /activate` _feels_ confusing. Would the account be reverted to a registered state?

NobodysNightmare commented:
I had an on-site discussion with @ulferts regarding the lock endpoint.

### GET lock

We would propose to drop the `GET lock` and instead include that information in the user resource (see below). This would be fine as the links from the user to the lock/unlock endpoints should suffice. Embedding the status in the user in some way seems neccessary (e.g. for index) and this makes the `GET lock` basically redundant. Consequently the return code of a successful lock/unlock could be changed to be an HTTP 200 and return the _affected user_ in its body. This seems like a good place where [returning 204 can be avoided](http://blog.ploeh.dk/2013/04/30/rest-lesson-learned-avoid-204-responses/).

What do you think?

### Idempotent transitions

We basically came to the same conclusion as you regarding idempotent transitions. They work fine from a clients perspective and telling the client that it tried locking a locked user adds little value.

### Representing the (lock) status on the user

We could not come to a conclusion here. One simple way that comes to mind is a simple boolean, e.g.

``` JSON
"isLocked": false
```

however this does not seem to scale well and does not tell the client that there is a semantic connection between attributes like `isLocked` and `isRegistered`. We discarded that approach.

It would probably make sense to represent the current user state as some kind of enumeration (this is what happens in the backend anyway).
We were left with two options still:

``` JSON
"status": "activated/locked/registered"
```

or 

``` JSON
"_links": {
    "status": { "href": "api/v3/user_statuses/activated" }
}
```

The former being a quite simple enum, where we can define a set of valid values and the latter being more complex in the first place, but offering options like:
- an index action to discover possible statuses
- further links from the status resource, e.g. listing users with this status

We are not quite sure if solution two is worth the additional effort or if it is plainly overengineered.

### Transitions from/to other statuses

We did not specifically talk about those yet. You are right that `DELETE activate` seems to be confusing, I'd not know what to expect from such a request, it should probably simply be invalid.

**IMO** it would suffice for lock/unlock to accept a user that is itself in active or locked state (idempotent as described above). Trying to lock/unlock a registered user should result in an error (probably HTTP 400 or 422 with a newly crafted `errorIdentifier`).

Thanks for the effort you ae putting into this! Best regards!

NobodysNightmare commented:
@ulferts and I talked again and decided that we want to stick to a simple enum on the topic of **representing the status on the user**. This should be sufficient for most use cases and is not as complex to implement.

I'd suggest simply having `"status": "active"`, where the value can be one of
- active
- registered
- locked

I hope that clears most uncertainties and allows you to continue to finish this specification. Tell me if something was left unclear. I tend to write too many words, so this might safely happen :wink:

oliverguenther committed:
Adding account status flag to user model, changed (un-)lock response

oliverguenther commented:
Missing the table was not intended, my bad. I've updated the PR with it and the latest changes from your discussion:
- Returning the user model with a 200 on lock, unlock
- Removing `GET /lock`
- added a "status": "active" to the model

Thanks for your efforts!

NobodysNightmare commented:
Great! The last thing missing from my perspective is the error handling in case the client is trying to lock/unlock a registered user. If a client tried to do that the response should probably (if you agree) be HTTP 400.

We should also assign a unique error identifier for that case as no identifier currently available matches that purpose. I'd suggest `urn:openproject-org:api:v3:errors:InvalidUserStatusTransition`.
See `## List of Error Identifiers` for the currently defined error identifiers.

oliverguenther committed:
Added InvalidUserStatusTransition to (un-)lock transitions

oliverguenther commented:
Added. I'm not certain to what extend the actual conflicting account states need to be documented in the response descriptions (or in the responses themselves, for that matter).

NobodysNightmare commented:
Looks good to me what you have done there :smile: 

I will probably add a word or two later, but for now it was only important to get a common understanding of possible error and success paths. That should enable you to write a proper implementation.

I will merge this PR now, as I believe we got everything in there. I am looking forward to your implementation :)

Best regards
Jan
